### PR TITLE
Rewrite crab spectrum as class

### PR DESCRIPTION
**Pull request by adonath**
This PR introduces a `CrabSpectrum` class, with the different reference models represented as `SpectrumModel`. This has the advantage, that all the methods of `SpectrumModel`, such as `.integral()`
and `energy_flux()` can be reused. It has now full `Quantity` support as well.

### CODE REVIEW

File: gammapy/spectrum/models.py [hunk 1]
```diff
@@ -167,6 +167,26 @@ def to_sherpa(self, name='default'):
         """
         raise NotImplementedError('{}'.format(self.__class__.__name__))
 
+    def spectral_index(self, energy):
+        """
+        Compute spectral index at given energy using a local powerlaw
+        approximisation.
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: approximisation -> approximation ?

File: gammapy/spectrum/crab.py [hunk 2]
```diff
@@ -15,138 +15,89 @@
...
-
+    Crab spectral model.
 
-def crab_spectral_index(energy=1, reference=CRAB_DEFAULT_REFERENCE):
-    """Spectral index (positive number) at a given energy.
+    The following references are available:
 
-    See the ``gammapy.spectrum.crab`` module docstring for a description
-    of the available reference spectra.
+        * 'meyer', 2010arXiv1008.4524M
```
Comment (cdeil) [inline gammapy/spectrum/crab.py]: No need to indent lists, putting them at the same level as the surrounding text gives nice HTML output.

File: gammapy/spectrum/models.py [hunk 3]
```diff
@@ -167,6 +167,26 @@ def to_sherpa(self, name='default'):
...
+        ----------
+        energy : `~astropy.units.Quantity`
+            Energy at which to estimate the index
+
+        Returns
+        -------
+        index : float
+            Estimated spectral index.
+        """
+        eps = np.sqrt(np.finfo(float).eps)
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: Does such a small `eps` yield numerically stable results?
I think I would have put something like `1e-3`.

But I don't know, if you think or know the spectral index with that small eps gives stable output, OK to keep as-is.

**Comment by cdeil** [comment]
After thinking about this some more, I'm thinking maybe keeping it a class it a good idea?
If we want to add extra methods like "convert given flux to Crab units" we could add it there and the method docstring would show up in the docs.

File: gammapy/spectrum/models.py (hunk 1 above)
Comment (adonath) [inline gammapy/spectrum/models.py]: Fixed

File: gammapy/spectrum/models.py (hunk 3 above)
Comment (adonath) [inline gammapy/spectrum/models.py]: I made it an option to the method. Default is now `1E-5`

File: gammapy/spectrum/crab.py (hunk 2 above)
Comment (adonath) [inline gammapy/spectrum/crab.py]: Fixed

**Comment by adonath** [comment]
@cdeil The use case with the unit is what I had in mind as well. Right now the following works, which isn't too bad:

```python
from astropy import units as u
from gammapy.spectrum import CrabSpectrum
from gammapy.spectrum.models import PowerLaw

# define power law model
amplitude = 1E-12 * u.Unit('1 / (cm2 s TeV)')
index = 2.3
reference = 1 * u.TeV
pwl = PowerLaw(index, amplitude, reference)

# define crab reference
crab = CrabSpectrum('hess_pl').model

# compute flux at 10 TeV in crab units
energy = 10 * u.TeV
flux_cu = (pwl(energy) / crab(energy)).to('%')
print(flux_cu)

# compute integral flux in crab units
emin, emax = [1, 10] * u.TeV
flux_int_cu = (pwl.integral(emin, emax) / crab.integral(emin, emax)).to('%')
print(flux_int_cu)
``` 

I think it's hard to implement a better API, because the flux unit is parametric and depends on what you do with the spectrum you want to measure the flux in crab units for...so something along the lines of:

```python
pwl(energy).to('crab')
pwl.integral(emin, emax).to('crab')
```
Is very hard to implement and probably not worth the effort.

Possibly one could have something like:
```python
pwl_cu = CrabSpectrum('meyer').use_as_unit_for(pwl)
pwl_cu(energy)
```
But it's not a super nice API either...

**Comment by cdeil** [comment]
I think what you have now, explaining how to do it via the docs, is a good solution.
Merge when travis-ci passes, right?

**Pull request merged**